triples-generator/src/main/scala/io/renku/triplesgenerator/events/consumers/tsmigrationrequest/migrations/CompositePlanProvision.scala

```diff
@@ -0,0 +1,184 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.renku.triplesgenerator.events.consumers.tsmigrationrequest.migrations
+
+import cats.Show
+import cats.data.{EitherT, OptionT}
+import cats.effect._
+import cats.kernel.Monoid
+import cats.syntax.all._
+import fs2.Stream
+import io.circe.Json
+import io.circe.syntax._
+import io.renku.compression.Zip
+import io.renku.events.{CategoryName, EventRequestContent}
+import io.renku.events.producers.EventSender
+import io.renku.graph.config.EventLogUrl
+import io.renku.graph.eventlog.EventLogClient
+import io.renku.graph.eventlog.EventLogClient.{EventPayload, SearchCriteria}
+import io.renku.graph.model.events.{EventInfo, EventStatus}
+import io.renku.graph.model.projects.{Path => ProjectPath}
+import io.renku.metrics.MetricsRegistry
+import io.renku.triplesgenerator.events.consumers.ProcessingRecoverableError
+import io.renku.triplesgenerator.events.consumers.ProcessingRecoverableError.LogWorthyRecoverableError
+import io.renku.triplesgenerator.events.consumers.tsmigrationrequest.Migration
+import io.renku.triplesgenerator.events.consumers.tsmigrationrequest.migrations.CompositePlanProvision.{Context, Result}
+import io.renku.triplesgenerator.events.consumers.tsmigrationrequest.migrations.tooling.{AllProjects, MigrationExecutionRegister, RecoverableErrorsRecovery, RegisteredMigration}
+import io.renku.triplesstore.SparqlQueryTimeRecorder
+import org.typelevel.log4cats.Logger
+
+private[migrations] class CompositePlanProvision[F[_]: Sync: Logger](
+    override val name: Migration.Name,
+    allProjects:       AllProjects[F],
+    eventSender:       EventSender[F],
+    eventLogClient:    EventLogClient[F],
+    executionRegister: MigrationExecutionRegister[F],
+    recoveryStrategy:  RecoverableErrorsRecovery = RecoverableErrorsRecovery
+) extends RegisteredMigration[F](name, executionRegister, recoveryStrategy) {
+
+  protected override def migrate(): EitherT[F, ProcessingRecoverableError, Unit] =
+    EitherT(
+      allProjects
+        .findAll(50)
+        .map(_.path)
+        .evalMap(findLatestEvent)
+        .unNone
+        .evalMap(CompositePlanProvision.createContext(getEventPayload))
+        .evalMap {
+          case Some(ctx) => sendStatusChange(ctx).as(Result.sendEvent)
+          case None      => Result.skip.pure[F]
+        }
+        .compile
+        .foldMonoid
+        .flatMap(result => Logger[F].info(result.show))
+        .attempt
+        .map(_.leftMap(LogWorthyRecoverableError(show"Error while running ts migration: $name", _)))
```

Comment:
Regarding errors, I wasn't quite sure. The only expected errors are when EL is down or its db migrations are running. This could theoretically happen in between. So when this would be restarted it should skip already done projects right? (because the latest event would be in state NEW). I think in general, it could be retried in this case?

Reply:
I think you could simply rely on `recoverWith(recoveryStrategy.maybeRecoverableError[F, Unit])`. It contains a common way of dealing with Exceptions.